autoload/phpactor/project/bufferMatcher.vim

```diff
@@ -0,0 +1,129 @@
+function! phpactor#project#bufferMatcher#assignFileToProject(filename)
+  if !exists('g:phpactorLoaded')
+    return
+  endif
+
+  if exists('b:project')
+    return
+  endif
+  let l:project = g:phpactorBufferMatcher.matchFileToProject(a:filename)
+  if v:null is l:project
+    return
+  endif
+  let b:project = l:project
+  if g:phpactorBufferMatcher.getRepository().hasProject(b:project)
+    return
+  endif
+
+  call g:phpactorBufferMatcher.getRepository().addProject(b:project)
+endfunction
+
+function! phpactor#project#bufferMatcher#create(projectRepository, projectRootMarkers, filesystemRootMarkers, initialCwd) abort
+  let l:filesystemRootMarkers = a:filesystemRootMarkers
+  if index(l:filesystemRootMarkers, '/') < 0
+    call add(l:filesystemRootMarkers, '/')
+  endif
+
+  return {
+        \ 'initialCwd': a:initialCwd,
+        \ 'projectRootMarkers': a:projectRootMarkers,
+        \ 'filesystemRootMarkers': l:filesystemRootMarkers,
+        \ 'repository': a:projectRepository,
+        \ 'getRepository': function('s:getRepository'),
+        \ 'matchFileToProject': function('s:matchFileToProject')
+        \ }
+endfunction
+
+function s:matchFileToProject(file) dict abort
+  let l:project = self.repository.findProjectContainingFile(a:file)
+
+  if type(l:project) == type({})
```

Comment:
What does `type({})` mean?

Reply:
It gave you the type number for a dictionary.
The constants `v:t_<type>` didn't always exists, fairly new even.
So that's how you can do it to be compatible with older versions:

Reply:
would it be better to check for `v:null`? `if (v:null !== l:project` ?

Reply:
Try `echo v:null != {}` (E735: Can only compare Dictionary with Dictionary). 

`isnot` can be used instead to compare with `v:null`. 

In some case of badly implemented function that should return `v:null` or a dictionary I got a number for example. Asking directly for dictionary type seems to be more accurate than comparing with `v:null` (although we can still get completely different object than we expect).

Reply:
What about `if !empty(l:project)`?

Reply:
Similarly as above: it can fail if potentially badly implemented `self.repository.findProjectContainingFile()` returned non-zero number for example.

Reply:
but seems that accepting any dict as the result is effectively as bad (it doesn't make us safer). Who will implement these functions? I think as a general approach I'd rather have `if something was returned` logic than `if returned value is of the expected type, return` (if nothing else because it's easier to understand) ?

Reply:
It's not like if someone else is going to implement this, I think the test here is not necessary.
The repository is an internal dictionary if we messed up the implementation of the function we don't care to have one more error here.
I would even say that I prefer to have as many error as possible in order to be sure to see them.
In PHP it would be a `LogicException`, you don't want to hide those otherwise you're not able to know that something went wrong.